nixos/modules/services/web-apps/pingvin-share.nix

```diff
@@ -0,0 +1,208 @@
+
+        port = mkOption {
+          type = types.port;
+          default = 3000;
+          example = 8000;
+          description = mdDoc ''
+            The port that the frontend service of Pingvin Share will listen to.
+          '';
+        };
+      };
+
+      nginx = {
+        enable = mkEnableOption (mdDoc ''
+          a Nginx reverse proxy for Pingvin Share.
+        '');
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+
+    users.groups = mkIf (cfg.group == "pingvin") {
+      pingvin = { };
+    };
+
+    users.users = mkIf (cfg.user == "pingvin") {
+      pingvin = {
+        group = cfg.group;
+        description = "Pingvin Share daemon user";
+        isSystemUser = true;
+      };
+    };
+
+    networking.firewall.allowedTCPPorts = mkIf cfg.openFirewall [ cfg.frontend.port ];
+
+    systemd.services.pingvin-share-backend = {
+      description = "Backend service of Pingvin Share, a self-hosted file sharing platform.";
+
+      wantedBy = [ "multi-user.target" "pingvin-share-frontend.service" ];
+      wants = [ "network-online.target" ];
+      before = [ "pingvin-share-frontend.service" ];
+      after = [ "network.target" ];
+
+      environment = {
+        PRISMA_SCHEMA_ENGINE_BINARY = "${pkgs.prisma-engines}/bin/schema-engine";
+        PRISMA_QUERY_ENGINE_BINARY = "${pkgs.prisma-engines}/bin/query-engine";
+        PRISMA_QUERY_ENGINE_LIBRARY = "${pkgs.prisma-engines}/lib/libquery_engine.node";
+        PRISMA_INTROSPECTION_ENGINE_BINARY = "${pkgs.prisma-engines}/bin/introspection-engine";
+        PRISMA_FMT_BINARY = "${pkgs.prisma-engines}/bin/prisma-fmt";
+
+        PORT = toString cfg.backend.port;
+        DATABASE_URL = "file:${cfg.settings.dataDir}/pingvin-share.db?connection_limit=1";
+        DATA_DIRECTORY = cfg.settings.dataDir;
+      };
+
+      path = with pkgs; [
+        cfg.backend.package
+        openssl
+        prisma-engines
+      ];
+
+      script = ''
+        cd ${cfg.backend.package}
+        prisma migrate deploy
+        prisma db seed
+        ts-node dist/src/main
+      '';
+
+      serviceConfig = {
+        User = cfg.user;
+        Group = cfg.group;
+        Type = "simple";
+        Restart = "on-failure";
+        StateDirectory = mkIf (cfg.settings.dataDir == "/var/lib/pingvin-share") "pingvin-share";
+      };
+    };
+
+    systemd.services.pingvin-share-frontend = {
+      description = "Frontend service of Pingvin Share, a self-hosted file sharing platform.";
+
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" "pingvin-share-backend.service" ];
+      after = [ "network-online.target" "pingvin-share-backend.service" ];
+
+      environment = {
+        PORT = toString cfg.frontend.port;
+        API_URL = "${if cfg.settings.https then "https" else "http"}://${cfg.settings.hostname}";
+      };
+      path = with pkgs; [ cfg.frontend.package ];
+
+      script = ''
+        cd ${cfg.frontend.package}
+        next start
```

Comment:
I think something like this should work right?
- `serviceConfig.WorkingDirectory = cfg.frontend.package`
- `serviceConfig.ExecStart = "${cfg.frontend.package}/bin/next start"`

We don't want `bash` to be process systemd monitors.

Reply:
Done